wasm2c/wasm-rt-impl-mem.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2018 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This file is not meant to be used directly, and is only used by including it
+// in wasm-rt-impl.c. This file may be included multiple times to implement
+// memory operations on different memory tyeps --- wasm_rt_memory_t and
+// wasm_rt_memory_shared_t depending on the value of WASM_RT_SHARED_MEM_OPS.
+
+#include <wasm-rt-impl.h>
+
+#ifndef WASM_RT_SHARED_MEM_OPS
```

Comment:
I like how this shares code between the non-shared and shared versions of allocate/grow/free memory, but the multiply-included `.c` file seems a bit confusing. How do you feel about having the .c file use the preprocessor macros directly to define the non-shared operations and then conditionally the shared ones, i.e. the same strategy that wasm-rt-impl.c uses for `DEFINE_TABLE_OPS(funcref)` and `DEFINE_TABLE_OPS(externref)`?

Reply:
At a high level I don't mind either approach. In practice, the challenge with pre-processor macros is that they aren't easy to work with while debugging --- single step is generally broken when dealing with pre-processor macros. The nice bit about this double include approach is that single step debugging works flawlessly without an issues or extra setup. Given this my first inclination would be to leave it as is, but I am not wedded to the idea. Let me know your thoughts.

Reply:
I guess my preference would be to keep the style consistent (e.g. with what we do for tables). I guess if you want this single-steppable in a debugger (but to reuse the shared part of these routines in the human-written source code), we could think about using a code generator or something?

Reply:
Ok. I will work on modifying this to use a code-gen style. I'll probably try to follow a similar format as used in the `src/template` folder.